File: javaNode/src/main/java/org/kevoree/library/command/AddDeployUnit.java
package org.kevoree.library.command;

import org.kevoree.DeployUnit;
import org.kevoree.KevoreeCoreException;
import org.kevoree.adaptation.AdaptationCommand;
import org.kevoree.adaptation.AdaptationType;
import org.kevoree.adaptation.KevoreeAdaptationException;
import org.kevoree.service.RuntimeService;
import org.kevoree.modeling.api.KMFContainer;

/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 26/01/12
 * Time: 16:35
 */

public class AddDeployUnit implements AdaptationCommand {

    private DeployUnit du;
    private RuntimeService runtimeService;

    public AddDeployUnit(DeployUnit du, RuntimeService runtimeService) {
        this.du = du;
        this.runtimeService = runtimeService;
    }

    @Override
    public void execute() throws KevoreeAdaptationException {
        ClassLoader classLoader;
        try {
            classLoader = runtimeService.installDeployUnit(du);
            if (classLoader == null) {
                throw new KevoreeAdaptationException("Unable to add DeployUnit " + du.path());
            }
        } catch (KevoreeCoreException e) {
            throw new KevoreeAdaptationException("Unable to add DeployUnit " + du.path(), e);
        }
    }

    @Override
    public void undo() throws KevoreeAdaptationException {
        new RemoveDeployUnit(du, runtimeService).execute();
    }

    @Override
    public AdaptationType getType() {
        return AdaptationType.ADD_DEPLOYUNIT;
    }

    @Override
    public int hashCode() {
        return getType().hashCode() + du.path().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof AdaptationCommand && obj.hashCode() == hashCode();
    }

    @Override
    public String toString() {
        return "AddDeployUnit    " + du.path();
    }

    @Override
    public KMFContainer getElement() {
        return du;
    }
}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/defaultNodeTypes/command/AddDeployUnit.kt
package org.kevoree.library.defaultNodeTypes.command

import org.kevoree.library.defaultNodeTypes.ModelRegistry
import org.kevoree.DeployUnit
import org.kevoree.api.PrimitiveCommand
import org.kevoree.log.Log

/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 26/01/12
 * Time: 16:35
 */

class AddDeployUnit(val du: DeployUnit, val bs: org.kevoree.api.BootstrapService, val registry: ModelRegistry) : PrimitiveCommand {

    override fun undo() {
        RemoveDeployUnit(du, bs, registry).execute()
    }

    override fun execute(): Boolean {
        try {
            var kclResolved = bs.get(du)
            if (kclResolved == null) {
                val new_kcl = bs.installDeployUnit(du)
                if (new_kcl != null) {
                    registry.register(du, new_kcl)
                    return true
                } else {
                    return false
                }
            } else {
                if(registry.lookup(du) == null){
                    registry.register(du, kclResolved)
                }
                return true
            }
        } catch(e: Exception) {
            Log.debug("error ", e); return false
        }
    }

    fun toString(): String {
        return "AddDeployUnit " + du.groupName + "/" + du.name + "/" + du.version + "/" + du.hashcode
    }


}